File: WTCChatApp/Models/Message.swift
//
//  Message.swift
//  WTCChatApp
//
//  Created by WTC Challenge
//

import Foundation

struct Message: Codable, Identifiable {
    let id: UUID
    var type: MessageType
    var senderId: String?
    var recipientId: UUID?
    var segmentTags: [String]?
    var content: MessageContent
    var status: String?
    var readAt: Date?
    var starred: Bool
    var createdAt: Date

    init(id: UUID = UUID(), type: MessageType, senderId: String? = nil, recipientId: UUID? = nil,
         segmentTags: [String]? = nil, content: MessageContent, status: String? = nil,
         readAt: Date? = nil, starred: Bool = false, createdAt: Date = Date()) {
        self.id = id
        self.type = type
        self.senderId = senderId
        self.recipientId = recipientId
        self.segmentTags = segmentTags
        self.content = content
        self.status = status
        self.readAt = readAt
        self.starred = starred
        self.createdAt = createdAt
    }

    var isRead: Bool {
        readAt != nil
    }
}

enum MessageType: String, Codable {
    case chat = "CHAT"
    case campaign = "CAMPAIGN"
}

struct MessageContent: Codable {
    var title: String
    var body: String
    var imageUrl: String?
    var buttons: [ActionButton]?

    init(title: String, body: String, imageUrl: String? = nil, buttons: [ActionButton]? = nil) {
        self.title = title
        self.body = body
        self.imageUrl = imageUrl
        self.buttons = buttons
    }
}

struct ActionButton: Codable, Identifiable {
    var id: UUID = UUID()
    var label: String
    var action: String

    enum CodingKeys: String, CodingKey {
        case label
        case action
    }

    init(label: String, action: String) {
        self.label = label
        self.action = action
    }
}

File: backend/src/main/java/com/wtc/chatapp/model/Message.java
package com.wtc.chatapp.model;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.Instant;
import java.util.List;
import java.util.UUID;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Document(collection = "messages")
public class Message {

    @Id
    @Builder.Default
    private String id = UUID.randomUUID().toString();

    @Builder.Default
    private MessageType type = MessageType.CHAT;

    private String senderId;

    @Indexed
    private String recipientId;

    private List<String> segmentTags;

    private MessageContent content;

    @Builder.Default
    private MessageStatus status = MessageStatus.SENT;

    private Instant readAt;

    @Builder.Default
    private boolean starred = false;

    @Builder.Default
    private Instant createdAt = Instant.now();
}
